stratosphere/loader/source/ldr_ro_manager.cpp

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (c) 2018-2019 Atmosphère-NX
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms and conditions of the GNU General Public License,
+ * version 2, as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stratosphere/ro.hpp>
+#include "ldr_ro_manager.hpp"
+
+namespace sts::ldr::ro {
+
+    namespace {
+
+        /* Convenience definitions. */
+        constexpr PinId InvalidPinId = {};
+        constexpr size_t ProcessCountMax = 0x40;
+        constexpr size_t ModuleCountMax = 0x20;
+
+        /* Types. */
+        struct ModuleInfo {
+            ldr::ModuleInfo info;
+            bool in_use;
+        };
+
+        struct ProcessInfo {
+            PinId pin_id;
+            u64 process_id;
+            ncm::TitleId title_id;
+            ncm::TitleLocation loc;
+            ModuleInfo modules[ModuleCountMax];
+            bool in_use;
+        };
+
+        /* Globals. */
+        ProcessInfo g_process_infos[ProcessCountMax];
+
+        /* Helpers. */
+        ProcessInfo *GetProcessInfo(PinId pin_id) {
+            for (size_t i = 0; i < ProcessCountMax; i++) {
```

Comment:
This (and several other loops) can be a range-based for loop. Same thing, but more concise and less typing:

```cpp
for (ProcessInfo& info : g_process_infos) {
    if (info.in_use && info.pin_id == pin_id) {
        return &info;
    }
}
```

Reply:
Yeah, I hear you on that.

I actually explicitly prefer the loop variable -- it gets optimized out in almost every meaningful cases, and it's nice to have stylistic consistency across loops (since some loops actually do need the explicit variable).